arch/riscv/kernel/irq.c

```diff
@@ -58,7 +58,7 @@ struct irq_chip riscv_irq_chip = {
 void __init init_IRQ(void)
 {
 	unsigned int irq;
-	for (irq = 0; irq < NR_IRQS; irq++)
+	for (irq = 0; irq < 2; irq++)
```

Comment:
At first, I left in the call to irq_set_chip_and_handler(), but then all
the IRQ numbers are in use and you cannot add an interrupt controller.

NR_IRQS specifies the total number of IRQs available. Drivers for interrupt
controllers will call (indirectly) irq_set_chip_and_handler() using unused
IRQ numbers.

I like your suggestion, but is there something more specific we can use
than "BASE"? These are the CPU interrupt numbers, if I understand the code
correctly.

Which leads me to my next question, which is how does the CPU receive an
interrupt from a peripheral device? That would be neither timer nor
software interrupt.

-Jamey

On Fri, Jan 22, 2016 at 2:32 AM Albert Ou <notifications@github.com> wrote:

> In arch/riscv/kernel/irq.c
> <https://github.com/riscv/riscv-linux/pull/30#discussion_r50508294>:
>
> > @@ -58,7 +58,7 @@ struct irq_chip riscv_irq_chip = {
> >  void __init init_IRQ(void)
> >  {
> >  	unsigned int irq;
> > -	for (irq = 0; irq < NR_IRQS; irq++)
> > +	for (irq = 0; irq < 2; irq++)
>
> It seems strange to me to increase NR_IRQS but not call
> irq_set_chip_and_handler() for the extra IRQ numbers.
>
> If this is somehow necessary, could we avoid the magic constant 2? A
> possible alternative:
>
> #define NR_IRQS_BASE    2
> #ifdef CONFIG_NR_IRQS_EXTRA
> #define NR_IRQS     (NR_IRQS_BASE + CONFIG_NR_IRQS_EXTRA)
> #else
> #define NR_IRQS     NR_IRQS_BASE
> #endif
>
> So the loop becomes for (irq = 0; irq < NR_IRQS_BASE; irq++).
> Please feel free to suggest a better way.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/riscv/riscv-linux/pull/30/files#r50508294>.
>